src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs

```diff
@@ -652,6 +653,56 @@ protected StatementSyntax GetStatementContainingInvocationToExtractedMethodWorke
             }
+
+            protected override async Task<SemanticDocument> UpdateMethodAfterGenerationAsync(
+                SemanticDocument originalDocument,
+                OperationStatus<IMethodSymbol> methodSymbolResult,
+                CancellationToken cancellationToken)
+            {
+                // Only need to update for nullable reference types in return
+                if (methodSymbolResult.Data.ReturnType.GetNullability() != NullableAnnotation.Annotated)
+                {
+                    return await base.UpdateMethodAfterGenerationAsync(originalDocument, methodSymbolResult, cancellationToken).ConfigureAwait(false);
+                }
+
+                var syntaxNode = originalDocument.Root.GetAnnotatedNodesAndTokens(MethodDefinitionAnnotation).FirstOrDefault().AsNode();
+                if (syntaxNode == null || !(syntaxNode is MethodDeclarationSyntax methodDeclaration))
+                {
+                    return await base.UpdateMethodAfterGenerationAsync(originalDocument, methodSymbolResult, cancellationToken).ConfigureAwait(false);
+                }
+
+                var semanticModel = originalDocument.SemanticModel;
+                var methodOperation = semanticModel.GetOperation(methodDeclaration, cancellationToken);
+
+                var returnOperations = methodOperation.DescendantsAndSelf().OfType<IReturnOperation>();
+
+                var aggregatedFlowState = NullableFlowState.NotNull;
+
+                foreach (var returnOperation in returnOperations)
+                {
+                    var returnTypeInfo = semanticModel.GetTypeInfo(returnOperation.ReturnedValue.Syntax, cancellationToken);
+                    if (returnTypeInfo.Nullability.FlowState == NullableFlowState.MaybeNull)
+                    {
+                        aggregatedFlowState = NullableFlowState.MaybeNull;
+                    }
+                }
+
+                if (aggregatedFlowState == NullableFlowState.MaybeNull)
+                {
+                    // Flow state shows that return is correctly nullable
+                    return await base.UpdateMethodAfterGenerationAsync(originalDocument, methodSymbolResult, cancellationToken).ConfigureAwait(false);
+                }
+
+                // Return type can be updated to not be null
+                var newType = methodSymbolResult.Data.ReturnType.WithNullability(NullableAnnotation.NotAnnotated);
+                var newMethodDeclaration = methodDeclaration.ReplaceNode(methodDeclaration.ReturnType, newType.GenerateTypeSyntax());
+
+                var oldRoot = await originalDocument.Document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var newRoot = oldRoot.ReplaceNode(methodDeclaration, newMethodDeclaration);
+
+                var newDocument = originalDocument.Document.WithSyntaxRoot(newRoot);
+                return await SemanticDocument.CreateAsync(newDocument, cancellationToken).ConfigureAwait(false);
```

Comment:
this fixing up "after the fact" seems super hinky and error prone.  is there a reason we can't generate it with the right nullability right from the start?

Reply:
So I tried to make this work in generating the right nullability annotation and I found it too difficult since we can modify the code by adding return statements. I ended up basically writing a predictive flow analyzer for all the cases we could have to deal with where we would inject a return statement.  I'm time boxing to the current design of fixing up after the fact, and I'll do my best to put a reasonable explanation as a comment here as to why it's this way. The only option I see is to accept the annotated nullability as desired and move on, but we asserted that we would like to be smarter than that. 

I'm very much open to ideas and discussion on this, but I think this is the most robust solution at this time based on my investigation.

Reply:
The [nullability project](https://github.com/dotnet/roslyn/projects/43) has a "IDE cleanup" column to track debt we created; this might be worth an entry there.